extractor/filesystem/os/apk/extractor.go

```diff
@@ -129,67 +126,94 @@ func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]
 	return inventory, err
 }
 
+func groupApkPackageLines(scanner *bufio.Scanner) ([][]string, error) {
+	var groups [][]string
+	var group []string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if line != "" {
+			group = append(group, line)
+			continue
+		}
+		if len(group) > 0 {
+			groups = append(groups, group)
+		}
+		group = make([]string, 0)
+	}
+
+	if len(group) > 0 {
+		groups = append(groups, group)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error while scanning: %w", err)
+	}
+
+	return groups, nil
+}
+
 func (e Extractor) extractFromInput(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
 	m, err := osrelease.GetOSRelease(input.FS)
 	if err != nil {
 		log.Errorf("osrelease.ParseOsRelease(): %v", err)
 	}
 
-	rd := textproto.NewReader(bufio.NewReader(input.Reader))
-	pkgs := []*extractor.Inventory{}
-	for eof := false; !eof; {
-		// Return if canceled or exceeding deadline.
-		if err := ctx.Err(); err != nil {
-			return pkgs, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
+	scanner := bufio.NewScanner(input.Reader)
+	packageGroups, err := groupApkPackageLines(scanner)
```

Comment:
This adds O(n) memory usage, plus a lot of allocations and cache misses. Given that we parse one group at a time I don't see why this is needed.

Reply:
Updated so that each group is processed one at a time.

Reply:
lgtm thanks